soweego/linker/evaluate.py

```diff
@@ -75,58 +134,93 @@ def _compute_performance(test_index, predictions, test_vectors_size):
+        test_score = grid_search.score(dataset[test_index], target[test_index])
+        result[f'test_{scoring}'].append(test_score)
+        best_model = grid_search.best_estimator_
+        result['best_models'].append(best_model)
+
+    return result
+
+
+def average_k_fold(classifier, catalog, entity, k, dir_io, **kwargs):
     predictions, precisions, recalls, fscores = None, [], [], []
     dataset, positive_samples_index = train.build_dataset(
         'training', catalog, entity, dir_io)
-    k_fold = StratifiedKFold(n_splits=k, shuffle=True)
-    # scikit's stratified k-fold no longer supports multi-label data representation.
-    # It expects a binary array instead, so build it based on the positive samples index
-    binary_target_variables = dataset.index.map(
-        lambda x: 1 if x in positive_samples_index else 0)
+    k_fold, binary_target_variables = utils.prepare_stratified_k_fold(
+        k, dataset, positive_samples_index)
 
     for train_index, test_index in k_fold.split(dataset, binary_target_variables):
         training, test = dataset.iloc[train_index], dataset.iloc[test_index]
 
-        model = workflow.init_model(classifier, binarize, training.shape[1])
+        model = _initialize(classifier, dataset, kwargs)
         model.fit(training, positive_samples_index & training.index)
-        
+
         preds = model.predict(test)
-        
+
         p, r, f, _ = _compute_performance(
             positive_samples_index & test.index, preds, len(test))
 
         if predictions is None:
             predictions = preds
         else:
             predictions |= preds
-        
+
         precisions.append(p)
         recalls.append(r)
         fscores.append(f)
 
     return predictions, mean(precisions), std(precisions), mean(recalls), std(recalls), mean(fscores), std(fscores)
 
 
-def single_k_fold(classifier, catalog, entity, binarize, dir_io, k=5):
+def _initialize(classifier, dataset, kwargs):
+    if classifier is constants.SINGLE_LAYER_PERCEPTRON:
+        model = workflow.init_model(classifier, dataset.shape[1], **kwargs)
+    else:
+        model = workflow.init_model(classifier, **kwargs)
+
+    LOGGER.info('Model initialized: %s', model)
+    return model
+
+
+def single_k_fold(classifier, catalog, entity, k, dir_io, **kwargs):
     predictions, test_set = None, []
     dataset, positive_samples_index = train.build_dataset(
         'training', catalog, entity, dir_io)
+    k_fold, binary_target_variables = utils.prepare_stratified_k_fold(
+        k, dataset, positive_samples_index)
 
-    k_fold = StratifiedKFold(n_splits=k, shuffle=True)
-    
-    # scikit's stratified k-fold no longer supports multi-label data representation.
-    # It expects a binary array instead, so build it based on the positive samples index
-    binary_target_variables = dataset.index.map(
-        lambda x: 1 if x in positive_samples_index else 0)
+    model = _initialize(classifier, dataset, kwargs)
```

Comment:
Model is being recreated in the for loop below and this specific one is not used. So I think it can be safely deleted.

```suggestion
    
```